core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ZooCache.java

```diff
@@ -554,4 +616,32 @@ public Optional<ServiceLockData> getLockData(ServiceLockPath path) {
     return ServiceLockData.parse(lockData);
   }
 
+  @SuppressWarnings("deprecation")
+  @Override
+  public void process(WatchedEvent event) {
+    // ZooCache implements the Watcher interface so that it can
+    // receive events from the ZooSesssion session watcher.
+    switch (event.getState()) {
+      case SyncConnected:
+        log.info("{} ZooKeeper connection established, re-establishing watchers; {}", cacheId,
+            event);
+        clear();
+        setupWatchers(pathsToWatch);
+        break;
```

Comment:
SyncConnected doesn't just happen when establishing a new session. It also happens when a session is migrated from one ZooKeeper server to another. This would reset cause redundant persistent recursive watchers to be added to the same session. That might be okay if ZooKeeper checks them to see if they've already been added, but it seems sketchy to assume that this is a new session.

Reply:
This is OBE, the code doesn't do this anymore.